File: runtime/rust/src/testing.rs
// testing extern — Rust native implementations

pub fn almide_rt_testing_assert_approx(a: f64, b: f64, epsilon: f64) {
    assert!((a - b).abs() < epsilon, "assert_approx failed: {} != {} (epsilon {})", a, b, epsilon);
}

pub fn almide_rt_testing_assert_contains(haystack: &str, needle: &str) {
    assert!(haystack.contains(needle), "assert_contains failed: {:?} does not contain {:?}", haystack, needle);
}

pub fn almide_rt_testing_assert_gt(a: i64, b: i64) {
    assert!(a > b, "assert_gt failed: {} is not > {}", a, b);
}

pub fn almide_rt_testing_assert_lt(a: i64, b: i64) {
    assert!(a < b, "assert_lt failed: {} is not < {}", a, b);
}

pub fn almide_rt_testing_assert_ok<T: std::fmt::Debug, E: std::fmt::Debug>(r: &Result<T, E>) {
    assert!(r.is_ok(), "assert_ok failed: got {:?}", r);
}

pub fn almide_rt_testing_assert_some<T: std::fmt::Debug>(o: &Option<T>) {
    assert!(o.is_some(), "assert_some failed: got None");
}

File: runtime/js/testing.js
const __almd_testing = {
  assert_throws(f, expected) {
    try { f(); throw new Error("__no_throw__"); }
    catch (e) {
      if (e.message === "__no_throw__") throw new Error("assert_throws: expected error '" + expected + "' but function succeeded");
      if (!e.message.includes(expected)) throw new Error("assert_throws: expected error containing '" + expected + "' but got '" + e.message + "'");
    }
  },
  assert_contains(haystack, needle) {
    if (!haystack.includes(needle)) throw new Error("assert_contains failed\n  expected to contain: \"" + needle + "\"\n  in: \"" + haystack + "\"");
  },
  assert_approx(a, b, tolerance) {
    if (Math.abs(a - b) > tolerance) throw new Error("assert_approx failed\n  left:  " + a + "\n  right: " + b + "\n  diff:  " + Math.abs(a - b) + " > tolerance " + tolerance);
  },
  assert_gt(a, b) {
    if (a <= b) throw new Error("assert_gt failed: " + a + " is not greater than " + b);
  },
  assert_lt(a, b) {
    if (a >= b) throw new Error("assert_lt failed: " + a + " is not less than " + b);
  },
  assert_some(opt) {
    if (opt === null || opt === undefined) throw new Error("assert_some failed: got none");
  },
  assert_ok(result) {
    if (result instanceof __Err) throw new Error("assert_ok failed: got err(" + result.message + ")");
  },
};
